client/table.go

```diff
@@ -0,0 +1,781 @@
+		return roachencoding.EncodeUvarint(b, v.Uint()), nil
+
+	case reflect.Float32, reflect.Float64:
+		return roachencoding.EncodeNumericFloat(b, v.Float()), nil
+
+	case reflect.String:
+		return roachencoding.EncodeBytes(b, []byte(v.String())), nil
+	}
+
+	return nil, fmt.Errorf("unable to encode key: %s", v)
+}
+
+// decodeTableKey decodes a single element of a table key from b, returning the
+// remaining (not yet decoded) bytes.
+func decodeTableKey(b []byte, v reflect.Value) ([]byte, error) {
+	switch t := v.Addr().Interface().(type) {
+	case *[]byte:
+		b, *t = roachencoding.DecodeBytes(b, nil)
+		return b, nil
+	case *string:
+		var r []byte
+		b, r = roachencoding.DecodeBytes(b, nil)
+		*t = string(r)
+		return b, nil
+	}
+
+	switch v.Kind() {
+	case reflect.Bool:
+		var i int64
+		b, i = roachencoding.DecodeVarint(b)
+		v.SetBool(i != 0)
+		return b, nil
+
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
+		var i int64
+		b, i = roachencoding.DecodeVarint(b)
+		v.SetInt(i)
+		return b, nil
+
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		var i uint64
+		b, i = roachencoding.DecodeUvarint(b)
+		v.SetUint(i)
+		return b, nil
+
+	case reflect.Float32, reflect.Float64:
+		var f float64
+		b, f = roachencoding.DecodeNumericFloat(b)
+		v.SetFloat(f)
+		return b, nil
+
+	case reflect.String:
+		var r []byte
+		b, r = roachencoding.DecodeBytes(b, nil)
+		v.SetString(string(r))
+		return b, nil
+	}
+
+	return nil, fmt.Errorf("unable to decode key: %s", v)
+}
+
+// encodePrimaryKey encodes a primary key for the table using the model object
+// v. It returns the encoded primary key.
+func (m *model) encodePrimaryKey(v reflect.Value) ([]byte, error) {
+	var key []byte
+	key = roachencoding.EncodeBytes(key, []byte(m.name))
+
+	for _, col := range m.primaryKey {
+		f, ok := m.fields[col]
+		if !ok {
+			return nil, fmt.Errorf("%s: unable to find field %s", m.name, col)
+		}
+		var err error
+		key, err = encodeTableKey(key, v.FieldByIndex(f.Index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return key, nil
+}
+
+// decodePrimaryKey decodes a primary key for the table into the model object
+// v. It returns the remaining (undecoded) bytes.
+func (m *model) decodePrimaryKey(key []byte, v reflect.Value) ([]byte, error) {
+	var name []byte
+	key, name = roachencoding.DecodeBytes(key, nil)
+	if string(name) != m.name {
+		return nil, fmt.Errorf("%s: unexpected table name: %s", m.name, name)
+	}
+
+	for _, col := range m.primaryKey {
+		f, ok := m.fields[col]
+		if !ok {
+			return nil, fmt.Errorf("%s: unable to find field %s", m.name, col)
+		}
+		var err error
+		key, err = decodeTableKey(key, v.FieldByIndex(f.Index))
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return key, nil
+}
+
+// encodeColumnKey encodes the column and appends it to primaryKey.
+func (m *model) encodeColumnKey(primaryKey []byte, column string) []byte {
```

Comment:
Hmm, the code started getting a lot more casts when trying to use `proto.Key` here. I'm going to leave as-is for now.